File: sdk/ios/Sources/MosaicSDK/DeviceLocale.swift
import Foundation

/// The one canonical form of a locale tag, shared by every Mosaic boundary that
/// compares locales.
///
/// `Locale.current.identifier` is an ICU identifier, not a language tag: a
/// device with a region override reports `en_US@rg=gbzzzz`, a POSIX environment
/// reports `en_US.UTF-8`, and Java's `Locale.toString` reports
/// `en_US_#u-rg-gbzzzz`. Hosts pass those straight through. Comparing them raw
/// made analytics events fail the closed codec, `application.locale` rules
/// evaluate to unknown, and localization catalog lookups miss the only catalog
/// that existed.
///
/// `canonicalTag` implements the protocol's canonical form verbatim — see the
/// "Localization and locale resolution" section of `docs/protocol/v0.2.md` and
/// the reference implementation `protocol/tools/locale-resolution-v0.2.mjs`.
/// Placement targeting and catalog lookup deliberately share it, with exactly
/// one ruled asymmetry: only catalog lookup recovers the leading language
/// subtag (`catalogTag`).
enum MosaicDeviceLocale {
  /// The device locale for an SDK context field that must satisfy a contract
  /// bound, or `nil` when the device reports nothing representable. An absent
  /// field beats a fabricated language.
  static var currentContextTag: String? { contextTag(currentIdentifier()) }

  /// The device locale for the Placement decision context.
  ///
  /// Canonical when it can be, otherwise the device's own value unchanged: the
  /// evaluator applies this same canonicalization and must stay free to
  /// classify an unusable locale as present-but-unknown rather than receive an
  /// absent field or a substituted language.
  static var currentTargetingLocale: String {
    let identifier = currentIdentifier()
    return canonicalTag(identifier) ?? identifier
  }

  private static func currentIdentifier() -> String {
    if #available(iOS 16.0, macOS 13.0, tvOS 16.0, watchOS 9.0, *) {
      return Locale.current.identifier(.bcp47)
    }
    return Locale.current.identifier
  }

  /// The canonical comparison form of one locale tag, or `nil` when nothing
  /// usable remains.
  ///
  /// The value is first cut at the first `@`, `.`, or `#` — the ICU keyword,
  /// POSIX charset, and Java extension markers. Underscores then become
  /// hyphens, empty subtags are dropped, the tag is truncated at the first
  /// singleton subtag (`-u-`, `-t-`, `-x-`), the language subtag is lowercased,
  /// a four-letter script subtag is title-cased, and a two-letter or
  /// three-digit region is uppercased.
  ///
  /// Callers that match against something — a catalog key, an authored operand —
  /// must use this and treat `nil` as "not comparable", never as a match.
  static func canonicalTag(_ identifier: String) -> String? {
    var subtags: [Substring] = []
    for subtag in preCut(identifier).split(separator: "-", omittingEmptySubsequences: true) {
      // A one-character subtag introduces an extension sequence; everything
      // from there on is outside the comparable identity of a locale.
      guard subtag.count > 1 else { break }
      subtags.append(subtag)
    }
    guard (1...8).contains(subtags.count),
      matches(subtags[0], "^[A-Za-z]{2,8}$"),
      subtags.dropFirst().allSatisfy({ matches($0, "^[A-Za-z0-9]{1,8}$") })
    else { return nil }
    return subtags.enumerated().map { index, subtag -> String in
      if index == 0 { return subtag.lowercased() }
      if matches(subtag, "^[A-Za-z]{4}$") {
        return subtag.prefix(1).uppercased() + subtag.dropFirst().lowercased()
      }
      if matches(subtag, "^(?:[A-Za-z]{2}|[0-9]{3})$") { return subtag.uppercased() }
      return subtag.lowercased()
    }.joined(separator: "-")
  }

  /// The requested locale as a localization catalog candidate.
  ///
  /// Catalog lookup — and only catalog lookup — recovers the leading language
  /// subtag when the whole tag cannot be canonicalized, so
  /// `en-US-verylongsubtag` still reaches the `en` catalog. Placement targeting
  /// must not: recovering there would change which users match a Rule, while
  /// here the chain would otherwise fall through to the document's own fallback
  /// and the worst outcome is a less specific translation of the author's copy.
  static func catalogTag(_ identifier: String) -> String? {
    if let canonical = canonicalTag(identifier) { return canonical }
    guard
      let leading = preCut(identifier)
        .split(separator: "-", omittingEmptySubsequences: false).first
    else { return nil }
    return canonicalTag(String(leading))
  }

  /// The canonical tag reduced until an SDK context field will accept it, or
  /// `nil` when it cannot be.
  ///
  /// Reduction here narrows a valid but over-long tag to its language; it never
  /// rescues an uncanonicalizable one.
  static func contextTag(_ identifier: String) -> String? {
    guard let tag = canonicalTag(identifier) else { return nil }
    if isContextValid(tag) { return tag }
    guard let language = tag.split(separator: "-").first.map(String.init),
      isContextValid(language)
    else { return nil }
    return language
  }

  /// Everything before the first ICU keyword (`@`), POSIX charset (`.`), or
  /// Java extension (`#`) marker, with underscores normalized to hyphens.
  private static func preCut(_ identifier: String) -> String {
    identifier
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .prefix(while: { $0 != "@" && $0 != "." && $0 != "#" })
      .replacingOccurrences(of: "_", with: "-")
  }

  /// The intersection of the analytics event context and Placement decision
  /// context locale rules; the analytics bound is the tighter of the two.
  private static func isContextValid(_ tag: String) -> Bool {
    (2...35).contains(tag.utf8.count) && matches(tag, "^[A-Za-z]{2,8}(?:-[A-Za-z0-9]{1,8})*$")
  }

  private static func matches(_ value: some StringProtocol, _ pattern: String) -> Bool {
    value.range(of: pattern, options: .regularExpression) != nil
  }
}

File: sdk/android/mosaic/src/main/kotlin/dev/mosaic/sdk/DeviceLocale.kt
package dev.mosaic.sdk

import java.util.Locale

/**
 * The canonical comparison form of a locale tag, and the device locale in that form.
 *
 * Protocol 0.3 and Placement Decision 1 share one rule, deliberately: Placement targeting and
 * localization catalog lookup must not disagree about what "the same locale" is. Underscores become
 * hyphens, empty subtags are dropped, the tag is truncated at the first singleton subtag, the
 * language is lowercased, a script subtag is title case, and a two-letter or three-digit region is
 * uppercased. See `docs/protocol/v0.3.md` ("Localization and locale resolution") and the reference
 * implementations `protocol/tools/locale-resolution-v0.3.mjs` and
 * `protocol/tools/placement-decision-validation-v1.mjs`.
 *
 * Truncation is what this exists for. `Locale.getDefault().toLanguageTag()` is well-formed BCP-47,
 * but Android reports regional preferences as Unicode extensions, so a device with a region
 * override or a non-default calendar reports `en-US-u-rg-gbzzzz` or
 * `en-US-u-ca-japanese-fw-mon-mu-celsius`. Those pass the analytics locale pattern — every subtag
 * is 1-8 alphanumerics — but they are not `en-US`: `application.locale` rules miss, catalog lookup
 * misses, and the longer forms exceed the analytics context's 35-byte bound, which makes the event
 * unencodable and loses it. Every locale the SDK sends or matches comes through here.
 */
internal object MosaicDeviceLocale {
    /**
     * The current device locale in canonical form, or `null` when the platform reports nothing
     * usable.
     *
     * Null, never a substituted tag. `application.locale` is optional in both the Placement
     * decision context and the analytics event context, and an absent locale makes every locale
     * condition evaluate UNKNOWN through the three-valued evaluator. Substituting a plausible tag
     * such as `en` would instead be a claim: a device whose locale Mosaic cannot read would be
     * targeted, bucketed, and reported as an English device, and `does_not_exist` — the Rule that
     * asks "did this host report a locale at all" — would silently stop matching it.
     *
     * Never recovers a language subtag from a tag that cannot be canonicalized either: Placement
     * targeting must not retarget a malformed tag onto its broader language, because that changes
     * which users match a Rule. Catalog lookup does recover — see [canonicalLookupOrNull].
     */
    val current: String? get() = canonicalOrNull(rawDeviceTag())

    /**
     * The current device locale for the analytics event context, whose 35-byte bound is tighter
     * than the canonical form allows, or `null` when nothing usable is reported. A canonical tag
     * past the bound degrades to its language subtag, because an event that cannot be encoded is an
     * event that is lost; that degradation is bounded by what the device actually reported and
     * never invents a locale. Targeting reads [current] instead, so this bound never moves anyone
     * between Rules.
     */
    val currentForEventContext: String? get() = boundedDeviceTag()

    /**
     * The canonical form of one language tag or legacy `Locale.toString()` identifier, or `null`
     * when nothing usable remains.
     *
     * Callers that must tell "the host gave me nothing usable" apart from "the host asked for
     * English" need that null. Localization resolution is one: an unusable request contributes no
     * candidate rather than selecting an `en` catalog the document may happen to declare.
     */
    fun canonicalOrNull(identifier: String): String? {
        val subtags = mutableListOf<String>()
        for (subtag in preCut(identifier).replace('_', '-').split('-')) {
            if (subtag.isEmpty()) continue
            // A one-character subtag opens an extension or private-use sequence (`-u-`, `-t-`,
            // `-x-`); everything from there on is device detail, not locale identity.
            if (subtag.length == 1) break
            subtags.add(subtag)
        }
        if (subtags.isEmpty() || subtags.size > 8) return null
        if (!LANGUAGE_PATTERN.matches(subtags.first())) return null
        if (subtags.drop(1).any { !SUBTAG_PATTERN.matches(it) }) return null
        return subtags.mapIndexed { index, subtag ->
            when {
                index == 0 -> subtag.lowercase(Locale.ROOT)
                SCRIPT_PATTERN.matches(subtag) ->
                    subtag.lowercase(Locale.ROOT).replaceFirstChar { it.titlecase(Locale.ROOT) }
                REGION_PATTERN.matches(subtag) -> subtag.uppercase(Locale.ROOT)
                else -> subtag.lowercase(Locale.ROOT)
            }
        }.joinToString("-")
    }

    /**
     * The requested locale as a **catalog lookup** candidate.
     *
     * Lookup recovers the leading language subtag when the whole tag cannot be canonicalized, so
     * `en-US-verylongsubtag` still reaches the `en` catalog. Placement targeting deliberately does
     * not: recovery there would change which users match a Rule, whereas here the chain would
     * otherwise fall through to the document's own fallback and the worst outcome of recovery is a
     * less specific translation of the author's own copy.
     */
    fun canonicalLookupOrNull(identifier: String): String? =
        canonicalOrNull(identifier)
            ?: canonicalOrNull(preCut(identifier).replace('_', '-').substringBefore('-'))

    /**
     * A host reports an ICU identifier, not a language tag. `en_US@rg=gbzzzz` (keyword),
     * `en_US.UTF-8` (POSIX charset), and `en_US_#u-rg-gbzzzz` (Java `Locale.toString`) all denote
     * `en-US`; without the cut the remainder fails the subtag grammar and the whole tag is unusable.
     */
    private fun preCut(identifier: String): String =
        identifier.trim().takeWhile { it != '@' && it != '.' && it != '#' }

    private fun rawDeviceTag(): String =
        runCatching { Locale.getDefault().toLanguageTag() }.getOrNull().orEmpty()

    private fun boundedDeviceTag(): String? {
        val canonical = canonicalOrNull(rawDeviceTag()) ?: return null
        // The analytics event context bounds the locale at 35 bytes on top of the pattern, and a
        // long variant chain can still exceed that after truncation. Degrading to the language
        // subtag the device itself reported keeps the event encodable instead of losing it to the
        // bound. If even that will not fit, the field is omitted rather than invented: `locale` is
        // optional in the event context, so an absent locale is reportable and a wrong one is not.
        return canonical.takeIf(::withinContextBound)
            ?: canonical.substringBefore('-').takeIf(::withinContextBound)
    }

    /**
     * The intersection of the analytics event context and Placement decision context locale rules;
     * the analytics bound is the tighter of the two.
     */
    private fun withinContextBound(tag: String): Boolean =
        tag.toByteArray(Charsets.UTF_8).size in 2..35

    private val LANGUAGE_PATTERN = Regex("^[A-Za-z]{2,8}$")
    private val SUBTAG_PATTERN = Regex("^[A-Za-z0-9]{1,8}$")
    private val SCRIPT_PATTERN = Regex("^[A-Za-z]{4}$")
    private val REGION_PATTERN = Regex("^(?:[A-Za-z]{2}|[0-9]{3})$")
}
